### ISSUE TITLE
Login with google for the second time : email Attribute cannot be updated

### ISSUE THREAD

**Issue by uchar** [aws-amplify/amplify-js#3526]
**Describe the bug**
I have the same problem as this [#issue](https://github.com/aws/amazon-cognito-auth-js/issues/159). If you make a user pool with required email field then the second google login attempt fails.
changing required fields after creating UserPool is not possible and deleting the UserPool and making a new one doesn't look like a good solution to me because by doing it I will lose all my user's data!

**To Reproduce**
1. Make a new UserPool with email attribute as required field
2. Enable Google federation in UserPool
3. Try to login for the second time using hosted ui
4. It should give you this error: `email Attribute cannot be updated.`

**Expected behavior**
When we use same email address it shouldn't update email attribute!

**Comment by powerful23** [comment]
@uchar Hi, did you map the `email` attributes from google to the Cognito User Pool. There is an section `Attribute mapping` under `Federation` in the Cognito User Pool console that you can do the mapping.

**Comment by uchar** [comment]
> @uchar Hi, did you map the `email` attributes from google to the Cognito User Pool? There is an section `Attribute mapping` under `Federation` in the Cognito User Pool console that you can do the mapping.

Yes I did, and It works fine for the first time but trying to sign in for the second time give me that error

**Comment by powerful23** [comment]
@uchar I tried to reproduce issue but couldn't. I have the email attribute as required in my user pool setting.  Is your email attributes mutable? There is a `Set attribute read and write permissions` secton under `App clients`.

**Comment by jcharlesworthuk** [comment]
I'm experiencing a similar issue, same setup as @uchar but I'm getting "Exception processing authorization code" the second time I sign in.

In answer to your question @powerful23 ues, the email is writable under App Clients.  In fact for me there is a message there that says **Required attributes are always writable**

**Comment by adrian-brennan** [comment]
I can reproduce the same issue. email is not mandatory for my pool, and is selected as readable and writable by my app client. First time login succeeds (and the email attribute on my Cognito user is populated), subsequent logins fail with 'user+attributes%3A+email%3A+Attribute+cannot+be+updated' error.

I worked around it by creating a custom attribute called 'emailaddress' and mapped the Google email attribute to that.

**Comment by vladistan** [comment]
Got the same problem.  I am not using amplify,  just regular OIDC client library.  

I get this error when using Google and SAML login through ADFS.   Also doesn't matter if it's one of the standard attributes or a custom one.  I tried to map 'name' attribute from google to customName,  and now I am getting an error message saying 'customName' could not be updated

**Comment by sammartinez** [comment]
@uchar Are you still experiencing this issue with later versions of Amplify? Please let us know

**Comment by AlexeyKhmelev** [comment]
Still have this issue. Terrible. User pool is already in production. Can't recreate it...

**Comment by rpstreef** [comment]
What's the resolution here because this is still happening

**Comment by pavanelthepumisc** [comment]
Any updates on this? I'm having exactly same issue.

**Comment by wadewomersley** [comment]
Happening here too.

**Comment by uchar** [comment]
It seems after two years of my report, you guys still don't fix it 
@sammartinez Actually, I moved away from Cognito a long time ago and can't test it anymore. It was a horrible experience with a lot of bugs and a lack of necessary features

**Comment by wadewomersley** [comment]
I updated my CF template and set the email to mutable: true and it does work fine for me now.

**Comment by fredski02** [comment]
I also experience this issue.

**Comment by JoacoChamo** [comment]
I am also having the same problem

**Comment by kewur** [comment]
> I updated my CF template and set the email to mutable: true and it does work fine for me now.

you can't actually do this for existing user pools, you'll be hit by "Existing schema attributes cannot be modified or deleted."

**Comment by jahanarun** [comment]
I also faced same issue and I have to nuke my User Pool to make the `email` schema as mutable.
My experience with Cognito has been poor so far. 
I won't be recommending Cognito to anyone, unless AWS makes it a priority.
Azure AD experience was much better.

**Comment by uchar** [comment]
@sammartinez Hi, I think you should reopen this issue, since it's still happening and not fixed after a couple of years

**Comment by abdallahshaban557** [comment]
Hi @uchar - can you please submit a new Github issue for us following our template so that we can investigate this? https://github.com/aws-amplify/amplify-js/issues/new?assignees=&labels=&template=1.bug_report.yaml

**Comment by uchar** [comment]
@abdallahshaban557 
As I said before, I don't use Cognito anymore, just saying this because this thread's notification is still popping up in my inventory. 
Also not sure why you need a new issue,I think this thread's descriptions are pretty obvious and easy to follow for investigation

**Comment by abdallahshaban557** [comment]
Just re-opened the issue @uchar  - apologies, I did not notice that a previous automation on our repo (which we have removed) had automatically closed your issue.

**Comment by nadetastic** [comment]
Hi @jahanarun following up on this issue. Could you share some info with me, by running the following command in your app directory: `amplify diagnose --send-report`  and sharing the ID it returns?

**Comment by nadetastic** [comment]
Hi @jahanarun @wadewomersley @fredski02 @JoacoChamo 

I have been testing for this issue, and I haven't been able to reproduce it yet. Could you confirm if this issue is still occurring? And if so which version of `aws-amplify` and are you generating the Cognito resource with Amplify CLI or manually?

**Comment by nadetastic** [comment]
Following up on this issue:

This can happen only when importing an existing Cognito/UserPoolClient resource that was not created with `amplify add auth`. During my investigation I found that this can happen when using CloudFormation or the SDK and not including the required email attribute as part of the [WriteAttributes](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-writeattributes) for your App Client.

As of now, using the amplify cli or AWS Cognito console to create the userpool client will set the required attribute to writable by default and and is a setting that cannot be modified.

Below is an example where `email` is required but `family_name` is not.
![Image](https://user-images.githubusercontent.com/87148442/211118735-f884bd5c-bfc1-4a94-9ff4-26dd340dcaca.png)

As a work around to resolve this, you can create a new app client with the same configuration (making sure you set the attribute to writable) and use that with your application. Will update the docs to include this callout.

**Comment by ershovio** [comment]
<img width="1295" alt="Screenshot 2023-01-19 at 19 26 11" src="https://user-images.githubusercontent.com/3596272/213497567-80cf4637-54c7-4b7d-be1b-72e546537930.png">
I have permissions for both read and write for name and email. At the same time I got `error_description=Invalid+user+attributes%3A+name%3A+Attribute+cannot+be+updated.%0Aemail%3A+Attribute+cannot+be+updated.%0A+&error=invalid_request` after second login using Google SSO

**Comment by ershovio** [comment]
@nadetastic  can you please take a look?

**Comment by ershovio** [comment]
```
attribute_mapping=aws_cognito.AttributeMapping(
                email=aws_cognito.ProviderAttribute.GOOGLE_EMAIL, fullname=aws_cognito.ProviderAttribute.GOOGLE_NAME
            ),
```

**Comment by ershovio** [comment]
I use AWS CDK with Python

**Comment by nadetastic** [comment]
Could you share a screen shot of how this looks like from the cognito console? @ershovio 

To get there:
1. Select the user pool from the list at - https://us-east-1.console.aws.amazon.com/cognito/v2/idp/user-pools (make sure to also choose the correct region)
2. Go to "Sign-in experience" tab
3. Under "Federated identity provider sign-in" select Google
4. View and share the "Attribute mapping" section

**Comment by ershovio** [comment]
<img width="1381" alt="Screenshot 2023-01-19 at 22 00 40" src="https://user-images.githubusercontent.com/3596272/213535817-ba72c4fb-00ba-4d49-8de3-48f5877e70cd.png">

**Comment by ershovio** [comment]
@nadetastic any updates on it?

**Comment by nadetastic** [comment]
Hi @ershovio 

I haven't been able to reproduce this issue with the same configuration you have. Since you have both email and name attributes set to Writable, you should be able to im curious what else could be causing this issue.

Note that if you setup you resource with the Amplify CLI, or through the Cognito console, you will not face this error, which leads me to believe that there might be something else with how the CDK is generating your resource.

I've discussed this with some folks from the Cognito team and are continuing that discussion. 

One thing I recommend is to try the work around mentioned on the [Amplify documentation](https://docs.amplify.aws/cli/auth/import/#import-an-existing-cognito-user-pool) regarding the use of existing Cognito resources. In short, create a new public app client within your user pool. and manually set the required attributes to writable and test with via a test application with this app client

**Comment by ershovio** [comment]
I needed to change the email and name as mutable in the CDK template - it solved the problem

**Comment by nadetastic** [comment]
@ershovio Glad you were able to get this resolved. I'll go ahead and close this issue

Thank you!

**Comment by anjanvb** [comment]
@nadetastic I have the following attributes set to write, but I still encounter this issue with OIDC authentication. I don't think this issue is fully resolved.  First login seems to be fine but subsequent logins seem to be an issue. I can even see external (OIDC) user in the userPool user list. Please, reopen this.

FYI @abdallahshaban557.

```
const standardCognitoWriteAttributes = {       
      email: true,
      familyName: true,
      givenName: true,
      locale: true,
      phoneNumber: true
    };
```

**Comment by abdallahshaban557** [comment]
Hi @anjanvb  - can you please create a new Github issue with the details we need to troubleshoot this?

**Comment by vicodinvic1** [comment]
got the same issue

**Comment by nadetastic** [comment]
Hi @anjanvb @vicodinvic1 i have reopened this issue to look into the problems you are facing.

Can you confirm how you configured your Cognito resources?

**Comment by annjawn** [comment]
Ok so I think the console is a bit confusing.

My userpool is being created with CDK [`cognito.UserPool`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.UserPool.html) and previously I had it setup with this

```
email:{
    required: true,
    mutable: false
}
```

But then i had this in the code

```
new cognito.ClientAttributes()
    .withStandardAttributes({       
      email: true,
      familyName: true,
      givenName: true,
      locale: true,
      phoneNumber: true
    });
```
and assigning the above to [`readAttributes:`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.ClientAttributes.html) for the UserPool while adding the client.

This caused the console to show that the `email` attribute on the client app is both `read` and `write`, whereas that turned out not to be true. So I ended up deleting the entire stack and re-deploying with changing `mutable` to `true`, and it worked.

I guess the confusion stems from the fact that standard attributes within `writeAttributes` (documented [here](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.ClientAttributes.html)) and `standardAttributes` (documented [here](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.StandardAttributes.html)) are sort of treated differently, so I am not sure if this is a CDK/Cfn or a Cognito issue.

**Comment by anjanvb** [comment]
@abdallahshaban557 @nadetastic  so I think the console is a bit confusing.

My userpool is being created with CDK [`cognito.UserPool`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.UserPool.html) and previously I had it setup with this

```
email:{
    required: true,
    mutable: false
}
```

But then i had this in the code

```
new cognito.ClientAttributes()
    .withStandardAttributes({       
      email: true,
      familyName: true,
      givenName: true,
      locale: true,
      phoneNumber: true
    });
```
and assigning the above to [`readAttributes:`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.ClientAttributes.html) for the UserPool while adding the client.

This caused the console to show that the `email` attribute on the client app is both `read` and `write`, whereas that turned out not to be true. So I ended up deleting the entire stack and re-deploying with changing `mutable` to `true`, and it worked.

I guess the confusion stems from the fact that standard attributes within `writeAttributes` (documented [here](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.ClientAttributes.html)) and `standardAttributes` (documented [here](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.StandardAttributes.html)) are sort of treated differently, so I am not sure if this is a CDK/Cfn or a Cognito issue.

**Comment by nadetastic** [comment]
@anjanvb Thanks for the follow up and clarification, I'm glad you were able to resolve this.

I believe the confusion comes from the use of the terms `"writable"` and `"mutable"`. Note that these are different and independent, in the way that `"mutable"` is specific to the **userPool** and `"writable"` is specific to the **userPoolClient**.

A scenario to help elaborate on this:

1. You have the attribute set to `"immutable"` on the userPool - meaning once it is set, it cannot be changed
2. Have the same attribute set to `"writable"` on the userPoolClient - meaning the userPoolClient can write to the userPool attribute.
3. Login the first time, and the userPoolClient writes to the userPool attribute, setting the value.
4. Login a second time, and the userPoolClient fails to write to the userPool attribute as it is `"immutable"` and cannot be changed.

This will result in a situation similar to what you and others have faced on this issue.

In short, be sure you are setting the attribute to both `"mutable"` on the userPool config and `"writable"` on the userPoolClient config.
#### References
- [UserPool Mutable definition](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_SchemaAttributeType.html#CognitoUserPools-Type-SchemaAttributeType-Mutable)
- [UserPoolClient Writable definition](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateUserPoolClient.html#CognitoUserPools-CreateUserPoolClient-request-WriteAttributes)
- [Mapping Attributes To Identity Provider](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-specifying-attribute-mapping.html#cognito-user-pools-specifying-attribute-mapping-requirements)

**Comment by nadetastic** [comment]
@anjanvb  I have updated the docs and included a callout that mentions the difference between mutable and writable attributes. I'll mark this issue as closed but let me know if you have any questions, thank you!

**Comment by silentlight** [comment]
Hi everyone. So in order to federated login to work I need to have email set to mutable for my user pool? I currently have user pool where email is immutable and login with Google fails on second attempt.

**Comment by danielmouras** [comment]
This issue really isn't difficult to understand. The correct behavior should be that when someone tries to log in again with Google or Apple, the system shouldn't attempt to overwrite the existing email, especially not when we're just dealing with a sign-in function. This would be a much simpler solution. Yet here I am, five years later, still facing the same problem.

It's absurd that this task has been closed so many times, especially when, five years later, using version v6 with a different method (signInWithRedirect), I'm still experiencing the same issue. It seems like all the recent Amazon layoffs were from the Amplify team, because there's absolutely no support.

**Comment by proper-gentleman** [comment]
So what is the status on this? It seems it's still happening.

```
http://localhost:3000/#error_description=user.email%3A+Attribute+cannot+be+updated.%0A+&state=PBhQV3LDT4Y2UaKWV4Tveh9poOxcRopX&error=invalid_request
```

**Comment by anasnadeemws** [comment]
any update on this?

It's still happening
http://localhost:3000/#error_description=user.email%3A+Attribute+cannot+be+updated.%0A+&state=sBsteJz6rlkxGXuFFzu3VdBsCGui4hJk&error=invalid_request

**Comment by gambin** [comment]
Perhaps this is not the best thread to report it, but I was facing the same nightmare in a TerraForm implementation. Solved that turning some implementations explicit, such as:

```
resource "aws_cognito_user_pool" "user_pool" {
  name = var.user_pool_name
  auto_verified_attributes = ["email"]
  alias_attributes         = ["preferred_username"]

  schema {
    name = "email"
    attribute_data_type = "String"
    required = true
    mutable = true
  }

  schema {
    name = "preferred_username"
    attribute_data_type = "String"
    required = true
    mutable = false
  }
}
```

```
resource "aws_cognito_user_pool_client" "user_pool_client" {
  user_pool_id = aws_cognito_user_pool.user_pool.id
  name = var.client_name

  read_attributes = ["sub", "preferred_username"]
...
 ```

By the way, hashicorp docs say something about the `write_attributes`, but it simply doesn't work, requiring some Macgyverism that works...

More details: 
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cognito_user_pool
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cognito_user_pool_client#write_attributes-1

**Comment by sumitlni** [comment]
The resolution for this issue would be better documentation from AWS. They should indicate that email should NOT be marked immutable during User Pool creation. If the schema being used to create user pool marks email as Mutable, this error does not occur. If the pool has been created already with email marked immutable, there is no "fix" other than switching to a new pool.
